File: job-2022-autumn/264.ugly-number-ii.js
/*
 * @lc app=leetcode id=264 lang=javascript
 *
 * [264] Ugly Number II
 */

// @lc code=start
/**
 * @param {number} n
 * @return {number}
 */
var nthUglyNumber = function (n) {
    const uglyNumbers = [];

    let list2CurrentIndex = 0;
    let list3CurrentIndex = 0;
    let list5CurrentIndex = 0;

    let list2NextValue = 1;
    let list3NextValue = 1;
    let list5NextValue = 1;

    while (uglyNumbers.length < n) {
        const minValue = Math.min(
            list2NextValue,
            list3NextValue,
            list5NextValue,
        );

        uglyNumbers.push(minValue);

        if (list2NextValue === minValue) {
            list2NextValue = uglyNumbers[list2CurrentIndex] * 2;
            list2CurrentIndex++;
        }
        if (list3NextValue === minValue) {
            list3NextValue = uglyNumbers[list3CurrentIndex] * 3;
            list3CurrentIndex++;
        }
        if (list5NextValue === minValue) {
            list5NextValue = uglyNumbers[list5CurrentIndex] * 5;
            list5CurrentIndex++;
        }
    }

    return uglyNumbers[n - 1];
};
// @lc code=end

nthUglyNumber(1690);

File: Daily/264.ugly-number-ii.cpp
/*
 * @lc app=leetcode id=264 lang=cpp
 *
 * [264] Ugly Number II
 */
#include <cinttypes>
#include <queue>
#include <unordered_set>
#include <vector>

// @lc code=start
class Solution {
 public:
  int nthUglyNumber(int n) {
    std::unordered_set<std::int64_t> uglyNumberSet = {1};
    std::priority_queue<std::int64_t, std::vector<std::int64_t>, std::greater<>>
        uglyNumbersHeap;
    uglyNumbersHeap.push(1);

    for (int i = 2; i <= n; i++) {
      const std::int64_t currentUglyNumber = uglyNumbersHeap.top();
      uglyNumbersHeap.pop();
      std::vector<std::int64_t> newUglyNumbers;
      if (!uglyNumberSet.contains(currentUglyNumber * 2)) {
        newUglyNumbers.push_back(currentUglyNumber * 2);
      }
      if (!uglyNumberSet.contains(currentUglyNumber * 3)) {
        newUglyNumbers.push_back(currentUglyNumber * 3);
      }
      if (!uglyNumberSet.contains(currentUglyNumber * 5)) {
        newUglyNumbers.push_back(currentUglyNumber * 5);
      }

      uglyNumberSet.merge(std::unordered_set<std::int64_t>(
          newUglyNumbers.begin(), newUglyNumbers.end()));
      for (const auto newUglyNumber : newUglyNumbers) {
        uglyNumbersHeap.push(newUglyNumber);
      }
    }

    return static_cast<int>(uglyNumbersHeap.top());
  }
};
// @lc code=end

int main() {
  Solution sol;
  sol.nthUglyNumber(10);
}
